### ISSUE TITLE
message when edit opens multiple items for a component

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/unison#2583]
This is a follow-up item to #2425, which has been partially addressed. The `edit` command now opens all terms/types that are part of a cycle when the user calls `edit` with one of them.

This might surprise users, and (even worse) they might just delete one of the parts of the component from their scratch file because they don't know why it's there.

It would be nice to have a message in the `ucm` output and/or the a comment in scratch file (maybe above each piece of the cycle?) that explains to the user that these should all be edited at once.

Any changes to the `ucm` output might be more visible/helpful after #2582.

**Comment by mitchellwrosen** [comment]
@tstat asked me today if, when updating some subset of a component, whether we could just update all of the members of the component.

I think that's a great idea, and would probably obviate the need to put more in the user's buffer than they asked for, as is done currently. 

@aryairani mentioned that we considered this before but I couldn't find the discussion.

**Comment by ceedubs** [comment]
This seems like it could be nice.

I'm a bit concerned about whether there would be problematic/surprising edge cases. For example say that `a`, `b`, and `c` are all part of a component. If I edit `a` and make it no longer reference `b` and `c`, does that mean that `b` and `c` are both updated to point to the standalone version of `a`? I suspect so, but we should probably have tests that make sure that `a` is no longer part of a component but `b` and `c` still are. I guess that this means that the logic would kick in when there is an update of a term that was part of a component _before_ the update, which doesn't necessarily mean that the version being saved is part of a component.

**Comment by aryairani** [comment]
@ceedubs I think this was the scenario @stew mentioned in conversation too. /cc @mitchellwrosen, @tstat

I guess we could issue a warning or something.  We should wizard-of-oz it and see what feels right.

**Comment by mitchellwrosen** [comment]
Interesting corner case, I had not considered that.

Would it make sense to just do something analogous to normal `update`, i.e. if the update is type-preserving, propagate it automatically (including references from its own component-mates), otherwise somehow indicate that there is outstanding work to do?

**Comment by aryairani** [comment]
It does seem probably fine and doable to me.  You'd have to load the whole component back up, and then re-hash-components it, then update all of them. I think it might just do the right thing.  Not sure how much output there should be.  Probably just the normal amount, whatever that ends up being.

**Comment by mitchellwrosen** [comment]
Just to reiterate Cody's comment above, it seems like when updating a term in a component of length N we may end up with anywhere between 1 and N new components afterwards, not just 1.

For example, consider a component where

- A refers to B
- B refers to C
- C refers to A

If A is updated and no longer refers to B, then

- C ought to be updated to refer to the new A, like a normal update
- B transitively ought to be updated to refer to the new C

And in the end we'll have 3 singleton components for the new A, B, C.

**Comment by aryairani** [comment]
`edit` no longer opens multiple items for a component, so 🏆
